sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,306 @@ import org.apache.spark.util.Utils
+        parser.getCurrentToken match {
+          case VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT =>
+            parser.getFloatValue
+
+          case VALUE_STRING =>
+            // Special case handling for NaN and Infinity.
+            val value = parser.getText
+            val lowerCaseValue = value.toLowerCase
+            if (lowerCaseValue.equals("nan") ||
+              lowerCaseValue.equals("infinity") ||
+              lowerCaseValue.equals("-infinity") ||
+              lowerCaseValue.equals("inf") ||
+              lowerCaseValue.equals("-inf")) {
+              value.toFloat
+            } else {
+              throw new SparkSQLJsonProcessingException(s"Cannot parse $value as FloatType.")
+            }
 
-      case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT, DoubleType) =>
-        parser.getDoubleValue
-
-      case (VALUE_STRING, DoubleType) =>
-        // Special case handling for NaN and Infinity.
-        val value = parser.getText
-        val lowerCaseValue = value.toLowerCase()
-        if (lowerCaseValue.equals("nan") ||
-          lowerCaseValue.equals("infinity") ||
-          lowerCaseValue.equals("-infinity") ||
-          lowerCaseValue.equals("inf") ||
-          lowerCaseValue.equals("-inf")) {
-          value.toDouble
-        } else {
-          throw new SparkSQLJsonProcessingException(s"Cannot parse $value as DoubleType.")
+          case _ => failedConversion(parser, dataType)
         }
+      }
 
-      case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT, dt: DecimalType) =>
-        Decimal(parser.getDecimalValue, dt.precision, dt.scale)
+    case DoubleType =>
+      (parser: JsonParser) => skipFieldNameTokenIfExists(parser) {
+        parser.getCurrentToken match {
+          case VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT =>
+            parser.getDoubleValue
+
+          case VALUE_STRING =>
+            // Special case handling for NaN and Infinity.
+            val value = parser.getText
+            val lowerCaseValue = value.toLowerCase
+            if (lowerCaseValue.equals("nan") ||
+              lowerCaseValue.equals("infinity") ||
+              lowerCaseValue.equals("-infinity") ||
+              lowerCaseValue.equals("inf") ||
+              lowerCaseValue.equals("-inf")) {
+              value.toDouble
+            } else {
+              throw new SparkSQLJsonProcessingException(s"Cannot parse $value as DoubleType.")
+            }
 
-      case (VALUE_NUMBER_INT, ByteType) =>
-        parser.getByteValue
+          case _ => failedConversion(parser, dataType)
+        }
+      }
 
-      case (VALUE_NUMBER_INT, ShortType) =>
-        parser.getShortValue
+    case StringType =>
+      (parser: JsonParser) => skipFieldNameTokenIfExists(parser) {
+        parser.getCurrentToken match {
+          case VALUE_STRING =>
+            UTF8String.fromString(parser.getText)
 
-      case (VALUE_NUMBER_INT, IntegerType) =>
-        parser.getIntValue
+          case token if token != VALUE_NULL =>
```

Comment:
Oh, yes this was a bit dirty one and related your comment above, https://github.com/apache/spark/pull/14102#discussion_r71096388.

It was also related with it was [here](https://github.com/apache/spark/blob/a95252823e09939b654dd425db38dadc4100bc87/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala#L83-L84) and [here](https://github.com/apache/spark/blob/a95252823e09939b654dd425db38dadc4100bc87/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala#L93-L95). `VALUE_NULL` was always being checked first so, it did not have to check `token != VALUE_NULL` before [here]((https://github.com/apache/spark/blob/a95252823e09939b654dd425db38dadc4100bc87/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala#L119-L124)).

However, I move both `case null | VALUE_NULL =>` and `case _ if parser.getTextLength < 1 =>` to the last in `failedConversion` (they were being checked first). 

Should I make another function for both cases and make this check them first like the original one?